### PR TITLE
Opensea royalty fees enforcement

### LINKED ISSUE #315: Update TheUbiquityStick nft contract to follow up opensea enforcement
### Reference 
https://twitter.com/opensea/status/1589058770646491136

We need to upgrade UbiquiStick.sol to follow up opensea's enforcement to receive royalty fee.
Implementation should be straightforward. Here is the acceptance criteria

- [ ] Integrate opensea-filter-registry into UbiquiStick.sol
- [ ] Deploy new UbiquiStick on goerli network
- [ ] Setup royalty fee and address & make sure if royalty works properly.

### PR DESCRIPTION
**Pull request by mul1sh**
Resolves #315 

- [x] Integrate opensea-filter-registry into UbiquiStick.sol
- [x] Deploy new UbiquiStick on goerli network
- [ ] Setup royalty fee and address & make sure if royalty works properly.

### CODE REVIEW

**Comment by mul1sh** [comment]
@pavlovcik @0xcodercrane  This is ready for review.

A few things to note,
- The contract has been [deployed to goerli](https://goerli.etherscan.io/address/0xe127e571b46f4de359136b89753965edb26c6d98#code) as required.
- In my opinion the 3rd task `Setup royalty fee and address & make sure if royalty works properly.` is not necessary because the way the **Opensea registry enforces fees** is by banning contracts known to bypass the royalty fees a collection owner sets in Opensea i.e. the contracts specified [here](https://github.com/ProjectOpenSea/operator-filter-registry#filtered-addresses)
- So I have added a test function called `updateOperator` in `UbiquiStick.sol` where you can test this ban functionality.
- To test with this function, pass in any address you have  as `address operator` and `bool filtered` should be `true` if you want the enforcement to be done by the opensea contract and `false` is you don't want any royalty enforcement.
- If `true` you will not be able to  `approve` or `transfer` an NFT after mint from the contract using the banned address
- If false the you will be able to  `approve` or `transfer` an NFT after mint from the contract using the approved address
- Final the `updateOperator` function is only for testing purposes and I will remove it before the final merge to `development`

File: packages/contracts/src/ubiquistick/UbiquiStick.sol
Comment (cuddle-dev) [inline packages/contracts/src/ubiquistick/UbiquiStick.sol]: ok. I got your idea. please make sure it gets removed once we merge it.

**Review by cuddle-dev — CHANGES_REQUESTED** [review]
Would you please create test file for UbiquiStick.sol including `updateOperator` function you created. It looks like `updateOperator` [reverted](https://goerli.etherscan.io/tx/0xb31d2ebff463ae3aab39e265a1caa8e3e47be22eb764d57b60026ca3d77ee0d1) on goerli network. so we recommend you to give us a proof in forge test. Others look good

**Comment by mul1sh** [comment]
@0xcodercrane thanks for the feedback, added the tests as required. 

Also the Etherscan call failed because the Etherscan input fields don't support input of  `booleans` because they always get formatted as strings/bytes. Anyway I've removed the test function because it's not necessary now with the forge tests.

File: packages/contracts/test/ubiquistick/UbiquiStick.t.sol
Comment (cuddle-dev) [inline packages/contracts/test/ubiquistick/UbiquiStick.t.sol]: we don't have a revert string here?

**Review by cuddle-dev — CHANGES_REQUESTED** [review]
Lets add tests for other functions you created

**Comment by mul1sh** [comment]
> Lets add tests for other functions you created

Done 🙂

**Comment by 0x4007** [comment]
@mul1sh a new rule we just added a couple days ago is no more force pushing. It makes the review process slower because we can't just compare the previous time we reviewed compared to the current moment review (we have to review all the changes since the beginning of the PR basically.)

Also strangely it appears that CI didn't run on your latest commit. Perhaps this is an effect of force pushing? Does your latest commit pass? The easy answer could be to do an empty commit to invoke CI so that we know its ready for a review.

Screenshot 2022-12-07 at 04 59 43

**Review by 0x4007 — APPROVED** [review]
Nothing stands out to me as an issue, but I'm also not a Solidity dev.

File: packages/contracts/test/ubiquistick/UbiquiStick.t.sol
Comment (cuddle-dev) [inline packages/contracts/test/ubiquistick/UbiquiStick.t.sol:194]: same as above

File: packages/contracts/test/ubiquistick/UbiquiStick.t.sol
Comment (cuddle-dev) [inline packages/contracts/test/ubiquistick/UbiquiStick.t.sol:98]: Lets add a check for the result of transfer like this `assertEq(afterBalance - beforeBalance, 1)` and transfer event as well

File: packages/contracts/test/ubiquistick/UbiquiStick.t.sol
Comment (cuddle-dev) [inline packages/contracts/test/ubiquistick/UbiquiStick.t.sol]: same as above

**Review by cuddle-dev — CHANGES_REQUESTED** [review]
Most stuff looks good to me. Lets just add missing checks and wrap up soon.

**Comment by 0x4007** [comment]
Looks like @0xcodercrane just wants a couple of extra checks added. Everything looks good to me. Looks like you're very close to completing the bounty! Thanks @mul1sh !

**Comment by mul1sh** [comment]
> Looks like @0xcodercrane just wants a couple of extra checks added. Everything looks good to me. Looks like you're very close to completing the bounty! Thanks @mul1sh !

@pavlovcik @0xcodercrane Done 🙂

**Review by cuddle-dev — APPROVED** [review]

**Comment by cuddle-dev** [comment]
Looks good!

**Pull request merged**
